log_entry.go

```diff
@@ -0,0 +1,102 @@
+package spectator
+
+import (
+	"strconv"
+	"strings"
+	"time"
+)
+
+type LogEntry struct {
+	registry *Registry
+	start    int64
+	id       *Id
+}
+
+func (entry *LogEntry) SetStatusCode(code int) {
+	entry.id = entry.id.WithTag("http.status", strconv.Itoa(code))
+}
+
+func (entry *LogEntry) SetSuccess() {
+	extraTags := map[string]string{
+		"ipc.result": "success",
+		"ipc.status": "success",
+	}
+	entry.id = entry.id.WithTags(extraTags)
+}
+
+func (entry *LogEntry) SetError(err string) {
+	extraTags := map[string]string{
+		"ipc.result": "failure",
+		"ipc.status": err,
+	}
+	entry.id = entry.id.WithTags(extraTags)
+}
+
+func (entry *LogEntry) SetAttempt(attemptNumber int, final bool) {
+	extraTags := map[string]string{
+		"ipc.attempt":       attempt(attemptNumber),
+		"ipc.attempt.final": strconv.FormatBool(final),
+	}
+	entry.id = entry.id.WithTags(extraTags)
+}
+
+func (entry *LogEntry) Log() {
+	duration := entry.registry.Clock().Nanos() - entry.start
+	r := entry.registry
+	r.config.IpcTimerRecord(r, entry.id, time.Duration(duration))
+}
+
+func attempt(n int) string {
+	switch n {
+	case 0:
+		return "initial"
+	case 1:
+		return "second"
+	default:
+		return "third_up"
+	}
+}
+
+func pathFromUrl(url string) string {
+	if url == "" {
+		return "/"
+	}
+
+	protoEnd := strings.IndexByte(url, ':')
+	if protoEnd < 0 {
+		return url
+	}
+
+	protocolLen := len(url) - protoEnd
+	if protocolLen < 3 {
+		// does not have ://
+		return url
+	}
+
+	// find the path, skipping over protocol://
+	pathBegin := strings.IndexByte(url[protoEnd+3:], '/')
+	if pathBegin < 0 {
+		return "/"
+	}
+	pathBegin += protoEnd + 3
+
+	queryBegin := strings.IndexByte(url[pathBegin+1:], '?')
+	if queryBegin < 0 {
+		// no query component
+		return url[pathBegin:]
+	}
+	queryBegin += pathBegin + 1
+
+	return url[pathBegin:queryBegin]
+}
+
+func NewLogEntry(registry *Registry, method string, url string) *LogEntry {
+	tags := map[string]string{
+		"owner":        "spectator-go",
+		"ipc.endpoint": pathFromUrl(url),
```

Comment:
Unless there is a guarantee that the path is fixed it is dangerous to include as a tag. Is this helper just meant for internal usage?

Reply:
It is intended only for internal use